external/llvm-project/mlir/lib/Conversion/AsyncToGPU/AsyncToGPU.cpp

```diff
@@ -0,0 +1,345 @@
+  }
+
+  return llvm::None;
+}
+
+// Get target{gpu} attribute from called func
+static Optional<DictionaryAttr> getGPUTarget(async::LaunchOp op) {
+  auto func = getCalledFunc(op);
+  if (!func.hasValue() || func->getNumResults() != 0)
+    return llvm::None;
+
+  auto attr = (*func)->template getAttrOfType<ArrayAttr>("targets");
+  if (!attr)
+    return llvm::None;
+
+  for (auto targetAttr : attr.getValue()) {
+    auto dictAttr = targetAttr.cast<DictionaryAttr>();
+    auto type = dictAttr.get("type");
+    if (type && type.template cast<StringAttr>() == "gpu")
+      return dictAttr;
+  }
+  return llvm::None;
+}
+
+//===----------------------------------------------------------------------===//
+// Convert async.launch ops with 'gpu' target to gpu.launch_func ops with
+// required memory staging.
+//===----------------------------------------------------------------------===//
+
+namespace {
+class LaunchOpConversion : public OpConversionPattern<async::LaunchOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  Value makeWait(OpBuilder b, Location loc, ArrayRef<Value> deps = {}) const {
+    auto tokenType = b.getType<gpu::AsyncTokenType>();
+    return b.create<gpu::WaitOp>(loc, tokenType, deps).asyncToken();
+  }
+
+  Value moveMemory(OpBuilder b, Value opr, uint32_t fidx,
+                   func::AccessMode accessMode,
+                   llvm::SmallVector<Value> &copyBackOprs,
+                   llvm::SmallVector<Value, 8> &asyncDeps) const {
+    auto loc = opr.getLoc();
+    auto tokenType = b.getType<gpu::AsyncTokenType>();
+    auto oprAllocOp = opr.getDefiningOp<memref::AllocOp>();
+    if (oprAllocOp)
+      b.setInsertionPoint(oprAllocOp);
+
+    auto allocWait = makeWait(b, loc);
+    auto gpuMemType = opr.getType();
+    auto dst = b.create<gpu::AllocOp>(loc, gpuMemType, tokenType,
+                                      ValueRange{allocWait}, ValueRange{},
+                                      ValueRange{});
+    auto dstMem = dst.getResult(0);
+    auto dstToken = dst.getResult(1);
+    // if alloc, convert to gpu.alloc
+    if (oprAllocOp) {
+      // TODO(sjw): make sure accessors are all on the GPU
+      oprAllocOp->replaceAllUsesWith(ValueRange{dstMem});
+    } else {
+      if (func::isAccessModeRead(accessMode)) {
+        // else copy to device
+        auto asyncDeps = ValueRange{dstToken};
+        auto memcpyToken =
+            b.create<gpu::MemcpyOp>(loc, tokenType, asyncDeps, dstMem, opr);
+        dstToken = memcpyToken.getResult(0);
+      }
+      if (func::isAccessModeWrite(accessMode)) {
+        copyBackOprs[fidx] = dstMem;
+      }
+    }
+    asyncDeps.push_back(dstToken);
+    return dstMem;
+  }
+
+  LogicalResult matchAndRewrite(async::LaunchOp op, OpAdaptor adaptor,
+                                ConversionPatternRewriter &rw) const override {
+    Location loc = op.getLoc();
+    auto module = op->getParentOfType<ModuleOp>();
+    auto ctx = module.getContext();
```

Comment:
At the very least, `auto* ctx`, but really `MLIRContext *ctx`

Reply:
I see plenty of auto upstream, especially when it's obvious and only used to pass thru. But fine.

Reply:
The standard I've seen (which, well, there's definitely stuff that's gone in that doesn't meet this rule, as everywhere, but I've seen reviewers upholding it and I see their point) is to use `auto` in cases only where the type is already mentioned nearby or when it's a long monstrosity that isn't really important (the exact types of iterators, for example).

But things like `Location loc = ...` seem preferred over `auto loc = ` because it means less running out to the API docs to find the type of the thing.

Reply:
I get that. But generally 'loc' is only ever passed to core methods, the details are never needed in this code.

Reply:
But that's not exactly the point - being able to generally tell what type a thing is (or what type it's being used as) is what's useful. Sure, it's not always important - a lot of type-inference-happy languages will lean on editor support for that question - but C++ doesn't have that culture, and so spelling out types that aren't clear from context (or way too long, or both) is a reasonable request to make of developers.

And so saying `Location`, `MLIRContext *` and so on explicitly helps people work their way around a large core codebase.

Reply:
'auto' has been part of the language for 10 years now. I think it is useful when details are not needed such as pass thru variables and avoids massive updates when type names change. 

In any case, the requested changes have been made and the tests are finally green.

Reply:
On the other hand, the people we'll be sending this code to think (from what I can tell) that while `auto` can help readability in some cases (ex. iterator types, the exact types of lambdas, some iterator variables, etc.), using it all over the place hurts readability and probably makes it easier to miss a spot during refactorings.

And so the general preference for naming your types (esp. for stuff like `Location`  or `MLIRContext *`)

... speaking of, you probably want `auto*` at the least on `getContext()` so it doesn't accidentally copy the context or something.